Torch/Utils/SteamWorkshopTools/SteamWorkshopTools.cs

```diff
@@ -0,0 +1,283 @@
+                    result[publishedFileId] = new PublishedItemDetails()
+                    {
+                        PublishedFileId = publishedFileId,
+                        Views           = fileDetails.GetValueOrDefault<uint>("views"),
+                        Subscriptions   = fileDetails.GetValueOrDefault<uint>("subscriptions"),
+                        TimeUpdated     = DateTimeOffset.FromUnixTimeSeconds(fileDetails.GetValueOrDefault<long>("time_updated")).DateTime,
+                        TimeCreated     = DateTimeOffset.FromUnixTimeSeconds(fileDetails.GetValueOrDefault<long>("time_created")).DateTime,
+                        Description     = fileDetails.GetValueOrDefault<string>("description"),
+                        Title           = fileDetails.GetValueOrDefault<string>("title"),
+                        FileUrl         = fileDetails.GetValueOrDefault<string>("file_url"),
+                        FileSize        = fileDetails.GetValueOrDefault<long>("file_size"),
+                        FileName        = fileDetails.GetValueOrDefault<string>("filename"),
+                        ConsumerAppId   = fileDetails.GetValueOrDefault<ulong>("consumer_app_id"),
+                        CreatorAppId    = fileDetails.GetValueOrDefault<ulong>("creator_app_id"),
+                        Creator         = fileDetails.GetValueOrDefault<ulong>("creator"),
+                        Tags            = tags.ToArray()
+                    };
+                }
+                return result;
+            }
+        }
+
+        public async Task DownloadPublishedFile(PublishedItemDetails fileDetails, string dir, string name = null)
+        {
+            var fullPath = Path.Combine(dir, name);
+            if (name == null)
+                name = fileDetails.FileName;
+            var expectedSize = (fileDetails.FileSize == 0) ? -1 : fileDetails.FileSize;
+
+            using (var client = new WebClient())
+            {
+                try
+                {
+                    var downloadTask = client.DownloadFileTaskAsync(fileDetails.FileUrl, Path.Combine(dir, name));
+                    DateTime start = DateTime.Now;
+                    for (int i = 0; i < 30; i++)
+                    {
+                        await Task.Delay(1000);
+                        if (downloadTask.IsCompleted)
+                            break;
+                    }
+                    if ( !downloadTask.IsCompleted )
+                    {
+                        client.CancelAsync();
+                        throw new Exception("Timeout while attempting to downloading published workshop item!");
+                    }
+                    //var text = await client.DownloadStringTaskAsync(url);
+                    //File.WriteAllText(fullPath, text);
+                }
+                catch (Exception e)
+                {
+                    throw new Exception($"{e.Message} - url: {fileDetails.FileUrl}, path: {Path.Combine(dir, name)}");
```

Comment:
This breaks the stack trace. Log your error first and then do `throw;`

Reply:
Good point, will do.